### PR TITLE
VisitDaoTest created and first part of tests as well

### PR DESCRIPTION
**Pull request by bohdancvejn**
+@Repository in VisitDaoImpl

### CODE REVIEW

File: persistence/src/test/java/cz/muni/fi/pa165/mushrooms/dao/VisitDaoTest.java [hunk 1]
```diff
@@ -0,0 +1,159 @@
...
+import javax.validation.ConstraintViolationException;
+
+import java.time.LocalDate;
+import java.util.Date;
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.assertj.core.api.Assertions.assertThatThrownBy;
+
+/**
+ * @author
```
Comment (Lindar84) [inline persistence/src/test/java/cz/muni/fi/pa165/mushrooms/dao/VisitDaoTest.java]: incognito :)

File: persistence/src/test/java/cz/muni/fi/pa165/mushrooms/dao/VisitDaoTest.java [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+        assertThatThrownBy(() -> visitDao.create(null)).hasRootCauseExactlyInstanceOf(IllegalArgumentException.class);
+    }
+
+    @Test
+    public void create_visitWithNullHunter() throws Exception {
+        Forest testForest = new Forest();
+        testForest.setName("Test Forest");
+
+        MushroomHunter testHunter = new MushroomHunter();
+
```
Comment (Lindar84) [inline persistence/src/test/java/cz/muni/fi/pa165/mushrooms/dao/VisitDaoTest.java]: unused hunter

File: persistence/src/test/java/cz/muni/fi/pa165/mushrooms/dao/VisitDaoTest.java [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
+        testVisit.setForest(testForest);
+        testVisit.setDate(LocalDate.of(2017,2,5));
+        testVisit.setNote("note for the forest");
+
+        assertThatThrownBy(() -> visitDao.create(testVisit)).isInstanceOf(ConstraintViolationException.class);
+    }
+
+    @Test
+    public void create_visitWithNullForest() throws Exception {
+        Forest testForest = new Forest();
```
Comment (Lindar84) [inline persistence/src/test/java/cz/muni/fi/pa165/mushrooms/dao/VisitDaoTest.java]: unused forest

File: persistence/src/test/java/cz/muni/fi/pa165/mushrooms/dao/VisitDaoTest.java [hunk 4]
```diff
@@ -0,0 +1,159 @@
...
+        Visit testVisit = new Visit();
+        testVisit.setHunter(testHunter);
+        testVisit.setForest(testForest);
+        testVisit.setDate(LocalDate.of(2017,2,5));
+        testVisit.setNote("blabla");
+
+        em.persist(testVisit);
+
+        assertThat(visitDao.findAll().size()).isEqualTo(sizeBeforeCreate + 1);
+        System.out.println(visitDao.findAll().size());
```
Comment (Lindar84) [inline persistence/src/test/java/cz/muni/fi/pa165/mushrooms/dao/VisitDaoTest.java]: Is it an intention to print the size?

File: persistence/src/test/java/cz/muni/fi/pa165/mushrooms/dao/VisitDaoTest.java (hunk 3 above)
Comment (bohdancvejn) [reply persistence/src/test/java/cz/muni/fi/pa165/mushrooms/dao/VisitDaoTest.java]: deleted for next version

File: persistence/src/test/java/cz/muni/fi/pa165/mushrooms/dao/VisitDaoTest.java (hunk 4 above)
Comment (bohdancvejn) [reply persistence/src/test/java/cz/muni/fi/pa165/mushrooms/dao/VisitDaoTest.java]: :) no, deleted for next version

File: persistence/src/test/java/cz/muni/fi/pa165/mushrooms/dao/VisitDaoTest.java (hunk 2 above)
Comment (bohdancvejn) [reply persistence/src/test/java/cz/muni/fi/pa165/mushrooms/dao/VisitDaoTest.java]: deleted for next version

File: persistence/src/test/java/cz/muni/fi/pa165/mushrooms/dao/VisitDaoTest.java (hunk 1 above)
Comment (bohdancvejn) [reply persistence/src/test/java/cz/muni/fi/pa165/mushrooms/dao/VisitDaoTest.java]: Thanks, fixed in next version

File: persistence/src/test/java/cz/muni/fi/pa165/mushrooms/dao/VisitDaoTest.java [hunk 5]
```diff
@@ -151,9 +148,237 @@ public void create_validVisit() throws Exception {
...
+        visitDao.update(foundVisit);
+        em.flush();
+
+        assertThat(em.find(Visit.class, visit1.getId()).getDate()).isEqualTo(newDate);
+    }
+
+    @Test
+    public void update_validVisitForestName() {
+        String forestNewName = "Sexy forest";
+        Forest foudForest = em.find(Forest.class, forest1.getId());
```
Comment (Lindar84) [inline persistence/src/test/java/cz/muni/fi/pa165/mushrooms/dao/VisitDaoTest.java]: typo in the name of the attribute :)

File: persistence/src/test/java/cz/muni/fi/pa165/mushrooms/dao/VisitDaoTest.java [hunk 6]
```diff
@@ -151,9 +148,237 @@ public void create_validVisit() throws Exception {
...
+        Visit foundVisit = em.find(Visit.class, visit1.getId());
+
+        foudForest.setName(forestNewName);
+
```
Comment (Lindar84) [inline persistence/src/test/java/cz/muni/fi/pa165/mushrooms/dao/VisitDaoTest.java]: I don't know, but isn't a forest.update() missing? And then, wouldn't it be too many tests in one @Test (visit and forest too)?

File: persistence/src/test/java/cz/muni/fi/pa165/mushrooms/dao/VisitDaoTest.java [hunk 7]
```diff
@@ -151,9 +148,237 @@ public void create_validVisit() throws Exception {
...
+        foundVisit.setForest(foudForest);
+        visitDao.update(foundVisit);
+        em.flush();
+
+        assertThat(em.find(Visit.class, visit1.getId()).getForest().getName()).isEqualTo(forestNewName);
+    }
+
+    @Test
+    public void update_validVisitForestDescription() {
```
Comment (Lindar84) [inline persistence/src/test/java/cz/muni/fi/pa165/mushrooms/dao/VisitDaoTest.java]: Doesn't it check the same problem as the test before? Is relevant what is changed in the forest attribute? Isn't important only that the forest was changed? And the same in tests with a hunter...

File: persistence/src/test/java/cz/muni/fi/pa165/mushrooms/dao/VisitDaoTest.java [hunk 8]
```diff
@@ -151,9 +148,237 @@ public void create_validVisit() throws Exception {
...
+        assertThatThrownBy(() ->visitDao.update(null)).hasRootCauseInstanceOf(IllegalArgumentException.class);
+    }
+
+    @Test
+    public void update_visitWithNullHunter() {
+        Forest testForest = new Forest();
+        testForest.setName("Test Forest");
+
+        em.persist(testForest);
+
```
Comment (Lindar84) [inline persistence/src/test/java/cz/muni/fi/pa165/mushrooms/dao/VisitDaoTest.java]: Isn't it redundant, if you have created forest1 and forest2 in @Before? Why don't you use one of them?
The same with a hunter in the next test...

File: persistence/src/test/java/cz/muni/fi/pa165/mushrooms/dao/VisitDaoTest.java [hunk 9]
```diff
@@ -151,9 +148,237 @@ public void create_validVisit() throws Exception {
         em.persist(testVisit);
 
         assertThat(visitDao.findAll().size()).isEqualTo(sizeBeforeCreate + 1);
```
Comment (Lindar84) [inline persistence/src/test/java/cz/muni/fi/pa165/mushrooms/dao/VisitDaoTest.java]: Shouldn't be "FindAll() test" the first of all tests, when you use the function in other tests?

File: persistence/src/test/java/cz/muni/fi/pa165/mushrooms/dao/VisitDaoTest.java [hunk 10]
```diff
@@ -0,0 +1,384 @@
...
+        visit1 = new Visit();
+        visit1.setHunter(mushroomHunter1);
+        visit1.setForest(forest1);
+        visit1.setDate(LocalDate.of(2017,10,25));
+
+        visit2 = new Visit();
+        visit2.setHunter(mushroomHunter2);
+        visit2.setForest(forest2);
+        visit2.setDate(LocalDate.of(2017,10,30));
+
```
Comment (bkompis) [inline persistence/src/test/java/cz/muni/fi/pa165/mushrooms/dao/VisitDaoTest.java]: Visits should be added to Forest and MushroomHunter instances too (yes we are missing the appropriate methods) - should not be the focus of the test for now, though.

File: persistence/src/test/java/cz/muni/fi/pa165/mushrooms/dao/VisitDaoTest.java [hunk 11]
```diff
@@ -0,0 +1,384 @@
...
+    @Test
+    public void create_nullVisit() throws Exception {
+        assertThatThrownBy(() -> visitDao.create(null)).hasRootCauseExactlyInstanceOf(IllegalArgumentException.class);
+    }
+
+    @Test
+    public void create_visitWithNullHunter() throws Exception {
+        Forest testForest = new Forest();
+        testForest.setName("Test Forest");
```
Comment (bkompis) [inline persistence/src/test/java/cz/muni/fi/pa165/mushrooms/dao/VisitDaoTest.java]: unnecessary new object, use the Forests you prepared in @Before

File: persistence/src/test/java/cz/muni/fi/pa165/mushrooms/dao/VisitDaoTest.java (hunk 7 above)
Comment (bkompis) [reply persistence/src/test/java/cz/muni/fi/pa165/mushrooms/dao/VisitDaoTest.java]: Which attribute was changed is relevant and should be tested (even if it's repetitive and seems pointless). Depends on how diligent we want to be - in my opinion, tests here on lines 187-260 are too much. Not bad, but not a priority.

File: persistence/src/test/java/cz/muni/fi/pa165/mushrooms/dao/VisitDaoTest.java [hunk 12]
```diff
@@ -0,0 +1,384 @@
...
+        visitDao.update(foundVisit);
+        em.flush();
+
+        assertThat(em.find(Visit.class, visit1.getId()).getHunter().getSurname()).isEqualTo(hunterNewSurname);
+    }
+
+    @Test
+    public void update_validVisitHunterNickName() {
+        String hunterNickName = "The boy who survived";
+        MushroomHunter foudHunter = em.find(MushroomHunter.class, mushroomHunter1.getId());
```
Comment (bkompis) [inline persistence/src/test/java/cz/muni/fi/pa165/mushrooms/dao/VisitDaoTest.java]: typo - foudHunter

File: persistence/src/test/java/cz/muni/fi/pa165/mushrooms/dao/VisitDaoTest.java [hunk 13]
```diff
@@ -0,0 +1,384 @@
...
+    @Test
+    public void findById_validId() throws Exception {
+        Visit testForest = visitDao.findById(visit1.getId());
+        assertThat(testForest).isNotNull();
+        assertThat(testForest).isEqualTo(visit1);
+    }
+
+    @Test
+    public void findById_invalidId() throws Exception {
+        Long invalidID = new Long(64000);
```
Comment (bkompis) [inline persistence/src/test/java/cz/muni/fi/pa165/mushrooms/dao/VisitDaoTest.java]: "random ID" - works, but is fragile and dependent on the database implementation. visit1.getId()+visit2.getId() should be safer.

**Review by bkompis — CHANGES_REQUESTED** [review]
Use your setup from @Before.

File: persistence/src/test/java/cz/muni/fi/pa165/mushrooms/dao/VisitDaoTest.java (hunk 5 above)
Comment (bohdancvejn) [reply persistence/src/test/java/cz/muni/fi/pa165/mushrooms/dao/VisitDaoTest.java]: fixed, thanks

File: persistence/src/test/java/cz/muni/fi/pa165/mushrooms/dao/VisitDaoTest.java (hunk 7 above)
Comment (bohdancvejn) [reply persistence/src/test/java/cz/muni/fi/pa165/mushrooms/dao/VisitDaoTest.java]: ok, already erased from the code.

File: persistence/src/test/java/cz/muni/fi/pa165/mushrooms/dao/VisitDaoTest.java (hunk 9 above)
Comment (bohdancvejn) [reply persistence/src/test/java/cz/muni/fi/pa165/mushrooms/dao/VisitDaoTest.java]: ok, moved to the top of the "find" methods

File: persistence/src/test/java/cz/muni/fi/pa165/mushrooms/dao/VisitDaoTest.java (hunk 11 above)
Comment (bohdancvejn) [reply persistence/src/test/java/cz/muni/fi/pa165/mushrooms/dao/VisitDaoTest.java]: thanks, updated

File: persistence/src/test/java/cz/muni/fi/pa165/mushrooms/dao/VisitDaoTest.java (hunk 8 above)
Comment (bohdancvejn) [reply persistence/src/test/java/cz/muni/fi/pa165/mushrooms/dao/VisitDaoTest.java]: thanks, changed as well

File: persistence/src/test/java/cz/muni/fi/pa165/mushrooms/dao/VisitDaoTest.java (hunk 12 above)
Comment (bohdancvejn) [reply persistence/src/test/java/cz/muni/fi/pa165/mushrooms/dao/VisitDaoTest.java]: corrected :) thanks

File: persistence/src/test/java/cz/muni/fi/pa165/mushrooms/dao/VisitDaoTest.java (hunk 13 above)
Comment (bohdancvejn) [reply persistence/src/test/java/cz/muni/fi/pa165/mushrooms/dao/VisitDaoTest.java]: ok, changed in the code

**Review by bkompis — APPROVED** [review]
Might be missing some tests.

**Pull request merged**